streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamImpl.java

```diff
@@ -388,76 +392,139 @@ public void process(final ProcessorSupplier<K, V> processorSupplier, String... s
+                             windows,
+                             keySerde,
+                             thisValueSerde,
+                             otherValueSerde);
+        }
 
-        KStreamKStreamJoin<K, R, V, V1> joinThis = new KStreamKStreamJoin<>(otherWindow.name(), windows.before, windows.after, joiner, outer);
-        KStreamKStreamJoin<K, R, V1, V> joinOther = new KStreamKStreamJoin<>(thisWindow.name(), windows.before, windows.after, reverseJoiner(joiner), outer);
+        // Both streams need repartitioning. Repartition and join the repartitioned streams
+        ensureJoinableWith((KStreamImpl<K, V1>) other);
+        KStreamImpl<K, V> thisRepartitioned = this.repartitionStream(keySerde,
+                                                                     thisValueSerde,
+                                                                     true);
+        KStreamImpl<K, V1> otherRepartitioned = ((KStreamImpl<K, V1>) other)
+            .repartitionStream(keySerde,
+                               otherValueSerde,
+                               true);
 
-        KStreamPassThrough<K, R> joinMerge = new KStreamPassThrough<>();
+        return join.join(thisRepartitioned,
+                         otherRepartitioned,
+                         joiner,
+                         windows,
+                         keySerde,
+                         thisValueSerde,
+                         otherValueSerde);
+    }
 
-        String thisWindowStreamName = topology.newName(WINDOWED_NAME);
-        String otherWindowStreamName = topology.newName(WINDOWED_NAME);
-        String joinThisName = outer ? topology.newName(OUTERTHIS_NAME) : topology.newName(JOINTHIS_NAME);
-        String joinOtherName = outer ? topology.newName(OUTEROTHER_NAME) : topology.newName(JOINOTHER_NAME);
-        String joinMergeName = topology.newName(MERGE_NAME);
 
-        topology.addProcessor(thisWindowStreamName, thisWindowedStream, this.name);
-        topology.addProcessor(otherWindowStreamName, otherWindowedStream, ((KStreamImpl) other).name);
-        topology.addProcessor(joinThisName, joinThis, thisWindowStreamName);
-        topology.addProcessor(joinOtherName, joinOther, otherWindowStreamName);
-        topology.addProcessor(joinMergeName, joinMerge, joinThisName, joinOtherName);
-        topology.addStateStore(thisWindow, thisWindowStreamName, otherWindowStreamName);
-        topology.addStateStore(otherWindow, thisWindowStreamName, otherWindowStreamName);
+    /**
+     * Repartition a stream. This will is required on join operations occurring after
+     * an operation that changes the key, i.e, selectKey, map(..), flatMap(..).
+     * @param keySerde      Serdes for serializing the keys
+     * @param valSerde      Serdes for serilaizing the values
+     * @param copartition   true if we need to co-partition these streams
+     * @return
+     */
+    private KStreamImpl<K, V> repartitionStream(Serde<K> keySerde,
```

Comment:
Nit: rename to "repartitionForJoin"? Since we have the other repartitioning before aggregation in another class.